### PR TITLE
[Core] Adding __copy__ and __deepcopy_ of Parameters to python

### LINKED ISSUE #8988: [core] Kratos Parameters copy constructor builds a non-independent object, just a pointer to the same Parameters.
**Description**
For a copy constructor, I would expect the creation of a new object, completely independent from the provided one.
The current implementation
https://github.com/KratosMultiphysics/Kratos/blob/0dbba4eca7d08ffcc251116c1b683ca73ea49d77/kratos/sources/kratos_parameters.cpp#L231
just copies pointers, so we essentially have the same object with a different name. If I wanted that I could create my own pointer or a reference.
I would expect a behavior like the one coded in the `= operator`:
https://github.com/KratosMultiphysics/Kratos/blob/0dbba4eca7d08ffcc251116c1b683ca73ea49d77/kratos/sources/kratos_parameters.cpp#L249
From Python, it is quite hard to get a deepcopy of Kratos Parameters, unless we change the behavior of this constructor. Then it would be very easy.
Can we change this? 
@RiccardoRossi 
@loumalouomega 
@KratosMultiphysics/technical-committee

### PR DESCRIPTION
**Pull request by loumalouomega**
**Description**
Expose __copy__ and __deepcopy_ of Parameters to python. Fixes #8988

**Changelog**
- Expose __copy__ and __deepcopy_ of Parameters to python

### CODE REVIEW

**Comment by maceligueta** [comment]
'new'? Are we sure that Python will call 'delete' when destroying the variable? Maybe a shared pointer would be better, I don't know...

**Comment by loumalouomega** [comment]
As I said I wrote on the phone. Probably is much better as you say.

**Comment by loumalouomega** [comment]
Ping @maceligueta

**Comment by maceligueta** [comment]
@loumalouomega let me know if you need support with writing the examples

**Comment by loumalouomega** [comment]
Done @philbucher and @maceligueta

**Comment by maceligueta** [comment]
@farrufat-cimne take a look at this. Good stuff

File: kratos/python/add_kratos_parameters_to_python.cpp [hunk 1]
```diff
@@ -69,6 +69,10 @@ void  AddKratosParametersToPython(pybind11::module& m)
     .def(py::init<const std::string&>())
     .def(py::init<std::ifstream&>())
     .def(py::init<Parameters const&>())
+    .def("__copy__", [](const Parameters& self) {auto a = Kratos::make_shared<Parameters>(self); return a;})
+    .def("__copy__", [](const Parameters& self, py::object memo) {auto a = Kratos::make_shared<Parameters>(self); return a;})
+    .def("__deepcopy__", [](const Parameters& self) {auto a = Kratos::make_shared<Parameters>(); Parameters& r_a = *a;r_a = self; return a;})
+    .def("__deepcopy__", [](const Parameters& self, py::object memo) {auto a = Kratos::make_shared<Parameters>(); Parameters& r_a = *a;r_a = self; return a;})
```
Comment (philbucher) [inline kratos/python/add_kratos_parameters_to_python.cpp]: Why two times?

Comment (loumalouomega) [reply kratos/python/add_kratos_parameters_to_python.cpp]: Because is required by the copy library, which is the standard to use the copy in python

Comment (philbucher) [reply kratos/python/add_kratos_parameters_to_python.cpp]: ah so the second version (the one accepting a `py::object memo` object) is for the `copy` lib?

**Comment by loumalouomega** [comment]
Yep

File: kratos/python/add_kratos_parameters_to_python.cpp [hunk 2]
```diff
@@ -69,6 +69,10 @@ void  AddKratosParametersToPython(pybind11::module& m)
     .def(py::init<const std::string&>())
     .def(py::init<std::ifstream&>())
     .def(py::init<Parameters const&>())
+    .def("__copy__", [](const Parameters& self) {auto a = Kratos::make_shared<Parameters>(self); return a;})
+    .def("__copy__", [](const Parameters& self, py::object memo) {auto a = Kratos::make_shared<Parameters>(self); return a;})
```
Comment (philbucher) [inline kratos/python/add_kratos_parameters_to_python.cpp]:
```suggestion
    .def("__copy__", [](const Parameters& self, py::object memo) {auto a = Kratos::make_shared<Parameters>(self); return a;}) // used for the copy module
```

File: kratos/python/add_kratos_parameters_to_python.cpp (hunk 1 above)
Comment (philbucher) [reply kratos/python/add_kratos_parameters_to_python.cpp]:
```suggestion
    .def("__deepcopy__", [](const Parameters& self, py::object memo) {auto a = Kratos::make_shared<Parameters>(); Parameters& r_a = *a;r_a = self; return a;}) // used for the copy module
```

**Review by philbucher — COMMENTED** [review]
ok for me, thx for the explanations

@maceligueta please feel free to approve if it works for you

**Comment by maceligueta** [comment]
This PR is useful for me and @farrufat-cimne , approving. Thanks all. 
However, I still think that the copy constructor should take care of this 'deepcopy' task, despite the extensive use it currently has as it is (a pointer copy). I commented that in #8988.

**Review by maceligueta — APPROVED** [review]

**Pull request merged**